### DISCUSSION TITLE
Ctrlr in 60 lessons

### DISCUSSION THREAD

**Discussion by tomlem** [RomanKubiak/ctrlr#318] [General]
A year ago I started to be interested in Ctrlr because I have owned a Roland sound module for 30 years that I could only operate manually until then.
I soon realized that I had to learn on three fronts that were relatively unfamiliar to me: the musical instruments digital interface standard, the Ctrlr language and finally the lua language.
I already owned a book on MIDI, as well as an extensive manual for my sound module. I bought two books about Lua ('Lua 5.3 Reference Manual' and 'Programming in Lua' by Roberto Jerusalimschy).
But there was nothing about the Ctrlr language but the blogs and the demo panels. This was the greatest difficulty.

What really helped me at the start of the analysis of the blogs was the simple language of the humble 'human fly'. He managed to force a place on the forum for the very basic questions of the novice loser. His questions and the answers to them from the seasoned and sometimes remarkably humble professors in ctrlr who took the time to answer these questions with great patience were of great benefit to me.

I then started collecting related explanations from different Ctrlr blogs into text files that I called 'lessons from the blog'.
Today this collection consists of 60 lessons.
The first lessons are not always so coherent because I didn't understand too much of them myself. The insight came only gradually. The lessons are also likely to be far from complete and may contain outdated or even faulty or not on topic parts. They are copy paste fragments, collected from the ctrlr blogs.

For the benefit of those who think they can use them, I am posting them today in attachment. They consist of 2 folders, one with the 60 text files ('lessons') and one with the panels or images mentioned in this texts.

[60LessonsFromCtrlrBlog.zip](https://github.com/RomanKubiak/ctrlr/files/6634535/60LessonsFromCtrlrBlog.zip)

May I invite you to improve on one or some or all of these lessons, weed out the superfluous, the faults and outdated parts an complete them so you can post them with a version number higher than 1.0 and with a short explanation of the improvements. 
And this for the benefit of all of us.

Thomas

**Comment by dnaldoog** [comment]
@tomlem It's excellent. I am wondering how all of this could be collated into a WIKI documentation page (with panel examples, including Atom's DEMO panels)? I think @shooking was talking about this, but I have no idea how it's done. If that main Ctrlr site ever goes down, it would be useful. That site seems to be being infiltrated by spammers! 

Ah Human Fly! Where is he these days? Those were the days! You are right. His questions were always the questions I wanted to ask when starting out!

**Reply by bijlevel** [reply]
@tomlem excellent initiative! @dnaldoog I remember The Puppeteer offering to host this kind of stuff on his server/website: https://godlike.com.au/index.php?id=361

**Reply by damiensellier** [reply]
@tomlem there's so many scripts and tips you picked up I also have as favorites  in my browser :) 
A wiki with all those great resources would be the best way to me.

**Reply by Godlike-Productions** [reply]
I'm happy to host this if you want.  Just let me know.  I'm not sure if I'm prepared to build a wiki due to the sheer amount of spam that they attract (I've build extensive ones in the past that we eventually had to shut down due to the spam volume), but happy to build a resource and a series of web pages.  I can also give select people editing rights on certain pages without too much drama.

**Reply by dobo365** [reply]
I can also host Ctrlr documentation on my Sunny Synths web site.
In fact I have already planned a section for Ctrlr related info in my test web site.

I would suggest that we host the **same** content at different places to secure continuation if for whatever reason one of the web site stops.
A third place would be the Documentation area in the main Ctrlr site. @dasfaker can support us with that and maybe also provide the rights to some of us to maintain it.

**Reply by Godlike-Productions** [reply]
Sounds like a great idea.  I'll swing a link over to your site, next time I'm doing some editing

**Comment by dobo365** [comment]
Thx for this work samoth.
However, let me do some positive critics ;-)
1. I checked a few of the lessons and they are not really readable even for someone knowing quite well Ctrlr. The issue is that it is just a reflection of the exchange as conversation on a specific topic. So we should completely clean each topic and document things as a kind of Q&A with the answer being to the point with also an example being to the point (sometimes people are publishing very long scripts while just a few lines are related to the topic
2. many comments are done by people that didn't know Ctrlr when posting them. The main example is for me human fly that started from scratch but took the habit to just post each of his reflection as long as he was progressing in the understanding. While having a good intention to support documentation of Ctrlr, I think he spammed the old forum with a lot of un-useful comments except the last ones he made. If I was posting every single reflection I'm making when searching for stuff you would also be spammed, hahaha ;-)

So, what I mean by those examples is that we need to seriously clean each topic before posting in a Wiki/documentation.
Maybe that we could also have a simple mini workflow like:
- someone propose a documentation on a topic
- some of us review and make comments (people not reviewing are agreeing by default)
- after for example a delay of 1w, we agree to publish, of course leaving the space for further completion/modifications but where at least the first publication is somewhat clean and maybe verified

**Reply by subv0x** [reply]
hello chaps :-) 
@tomlem thanks for the kind words - and some good reading (apart from some of my bits ! hehe)
@dnaldoog hey wassup << this guy showed me some *very* useful stuff. totally unlocked a new level of use and gave me the ability to resolve problems creatively. and that's with 2 or 3 tips and walkthroughs - a little can go a long way. kept me busy for a bit anyway.

i'm here to lament the loss of the forum, hey roman.kubiak - it was a really useful part of Ctrlr 'practice' ... nay: essential: Ctrlr needs a forum. maybe we can squat a part of someone else's forum? :D 

i picked up Ctrlr again recently, so i'm tinkering with old projects, extracting the useful bits.

**Comment by subv0x** [comment]
@dobo365 - yeah fair point, it was spamming, technically - especially the early stages haha
cries of agony. found it quite hard to get my head around Lua, with no 'coding' experience.

edit: actually, no it wasn't spamming at all: i kept my rambling to my own D110 thread most of the time, and posted examples of what i was doing to share - i realised that most people are too wrapped up in their own projects to want to pollute their Roaming folder with even more Ctrlr project folders. but i'm not actually 'publishing' (although i understand what you mean): i'm just an ordinary user, having a chat. this might not be 'the done thing' in programming circles. i picked up on that too, and saw that a bit of discretion was more welcome than apparently random outbursts of frustration. remember that there is very little documentation on Ctrlr, until you understand where to look for the information (such as the JUCE library). coming from a background of use of Synthedit, there was a steep initial learning curve, with 'the Lua'. ...

**Comment by subv0x** [comment]
re: topic subject
there are some interesting issues new users need to know about, such as modulators jumping to the wrong positions when you reload a panel, and why that happens, ways to avoid it, things to be aware of. (it's a bug really: modulators not shedding uiTabs or uiGroup ownership/parenthood when removed or copied from a 'container'.) 

or just best safe ways of saving. 
i had nightmares with these two at the beginning, losing work. (still not quite sure what 'save' does'. is it current loaded file version? logically, yes, but i'm not sure, and always 'save as' just to be sure. traumatised behaviour.)

and then there are files that will just stop everything else from working - i've got one of those. there's some conflict somewhere, probably with another panel i have loaded. have to run that one on its own. 

or the times you have to empty/delete your Roaming/Ctrlr folder contents. (have forgotten why i had to do this)

**Comment by h-WAVES** [comment]
您好，您的邮件已经收到，无法立即回复您的邮件请谅解,我会在第一时间给您回复。
    Hello, your mail has received, can not immediately reply to your mail, please forgive me, I will give you a reply in the first time
